syntax = "proto3";

package herm.market_data;

// Venue contribution at a price level
message VenueContribution {
  string venue_name = 1;
  double quantity = 2;
}

// Price level in the order book
message PriceLevel {
  double price = 1;
  double quantity = 2;  // Total aggregated quantity
  repeated VenueContribution venues = 3;  // Which venues contribute (optional, for venue-aware queries)
}

// Order book snapshot or update
message OrderBook {
  string symbol = 1;  // e.g., "BTCUSDT"
  int64 timestamp_us = 2;  // Microseconds since epoch
  repeated PriceLevel bids = 3;  // Sorted descending by price
  repeated PriceLevel asks = 4;  // Sorted ascending by price
  map<string, int64> venue_timestamps = 5;  // venue_name -> last_update_timestamp_us
}

// Request type for market data streaming
enum RequestType {
  NONE = 0;         // Full order book
  BBO = 1;          // Best bid/offer only
  PRICE_BAND = 2;   // Price bands at specified basis points
  VOLUME_BAND = 3;  // Volume bands at specified notional volumes
}

// Unified market data streaming request
message StreamMarketDataRequest {
  string symbol = 1;  // e.g., "BTCUSDT"
  RequestType request_type = 2;
  repeated int32 basis_points = 3;  // For PRICE_BAND: e.g., [50, 100, 200, 500, 1000]
  repeated double notional_volumes = 4;  // For VOLUME_BAND: e.g., [1000000, 5000000, ...]
}

// Market data streaming service
service MarketDataService {
  // Unified streaming endpoint - returns OrderBook filtered/resampled based on request_type
  rpc StreamMarketData(StreamMarketDataRequest) returns (stream OrderBook);
}
